### PR TITLE
fix cycle dependency in libxml+python

### LINKED ISSUE #13637: ROOT + Python/libxml2: Maximum Recursion Depth
Trying to install `root` via
```bash
$ spack spec root
```
leads to
```
Input spec
--------------------------------
root

Concretized
--------------------------------
==> Error: maximum recursion depth exceeded
```

The problem is gone when using `root ~python`.

---

### Steps to reproduce the issue

```console
$ spack install root
```

### Platform and user environment

Please report your OS here:
```commandline
$ uname -a 
Linux axel-work 5.0.0-29-generic #31~18.04.1-Ubuntu SMP Thu Sep 12 18:29:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	Ubuntu 18.04.3 LTS
``` 

### Additional information

Spack version: `develop` as of ae6229dee20da04b2120dc97cdfae800c136e42a

Note 1: `root +x` (default) does not build (seen via `root ~xml`).
Note 2: `root ~x` does not build unless I also deactivate `~opengl` (default: `+opengl`).

### LINKED ISSUE #13846: Circular dependency in libxml2 causes recursion error when global variant set
Reproducer:

`spack spec hwloc ^libxml2+python` or replace `hwloc` with any other package that depends on `libxml2`

```
bash-4.2$ ./bin/spack spec conduit ^libxml2+python
Input spec
--------------------------------
conduit

Concretized
--------------------------------
==> Error: maximum recursion depth exceeded while calling a Python object
```
Core issue:
 `libxml2` depends on `python` which depends on `libxml2` through `gettext`. 

@becker33

### PR DESCRIPTION
**Pull request by becker33**
fixes #13846
fixes #13637 (EDIT: added)
closes #13689 (EDIT: added)

When libxml2 depends on python, it needs to ensure python doesn't depend on libxml2 (through gettext). 

This PR adds a variant to python to control its dependency on gettext by forwarding the libxml2 variant. Default `False` in the python package. Then libxml2+python can depend on python~libxml2, eliminating the cycle.

### CODE REVIEW

**Comment by becker33** [comment]
@mclarsen, please confirm that this resolves your problem

**Comment by scheibelp** [comment]
https://github.com/spack/spack/pull/13689 handles this, although it removes `gettext` as a dependency from python rather than making it conditional.

I think #13689 is simpler: we generally don't have variants which control how dependencies are built. Discussion in #13637 (in particular https://github.com/spack/spack/issues/13637#issuecomment-552025873) led me to think that removing the `gettext` dependency was the sensible way to proceed.

Like #13689, this also fixes #13637

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -67,6 +67,10 @@ class Python(AutotoolsPackage):
 
     extendable = True
 
+    # Variants to control dependency variants to avoid cyclical dependencies
+    variant('libxml2', default=False,
+            description='Use a gettext library build with libxml2')
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/python/package.py]: @becker33 

"build" -> "built"

Also, if we go this route, then I think the associated comment needs more detail (and/or a reference to an issue/PR thread which explains it in more detail).

File: var/spack/repos/builtin/packages/libxml2/package.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ class Libxml2(AutotoolsPackage):
     depends_on('zlib')
     depends_on('xz')
 
-    depends_on('python+shared', when='+python')
+    depends_on('python+shared~libxml2', when='+python')
```
Comment (ax3l) [inline var/spack/repos/builtin/packages/libxml2/package.py]: I would also add a comment/reference before this line warning about the cyclic dependency

File: var/spack/repos/builtin/packages/python/package.py (hunk 1 above)
Comment (ax3l) [reply var/spack/repos/builtin/packages/python/package.py]: does Python even need libxml support in gettext or can we just decide to generally take the variant `gettext ~libxml` as dependency?

**Pull request merged**